marshal/claim.go

```diff
@@ -360,16 +367,36 @@ func (c *claim) messagePump() {
 		// needs to be serialized with Release, otherwise a race-condition can potentially
 		// lead to a write to a closed-channel. That's why we're using this lock. We're not
 		// using the main lock to avoid deadlocks since the write to the channel is blocking
-		// until someone consumes the message blocking all Commit operations
-		c.messagesLock.Lock()
-		if !c.Terminated() {
-			// This allocates a new Message to put the proto.Message in.
-			// TODO: This is really annoying and probably stupidly inefficient, is there any
-			// way to do this better?
-			tmp := Message(*msg)
-			c.messages <- &tmp
+		// until someone consumes the message blocking all Commit operations.
+		//
+		// This must not block -- if we hold the messagesLock for too long we will cause
+		// possible deadlocks.
+		//
+		// This allocates a new Message to put the proto.Message in.
+		// TODO: This is really annoying and probably stupidly inefficient, is there any
+		// way to do this better?
+		tmp := Message(*msg)
+		messagePending := true
+		for messagePending && !c.Terminated() {
+			messagePending = func() bool {
+				c.messagesLock.Lock()
+				defer c.messagesLock.Unlock()
+
+				select {
+				case c.messages <- &tmp:
+					return false // no longer pending
+				case <-time.After(10 * time.Millisecond):
```

Comment:
something that I've seen work really well in our code is to create a "stopChan" associated with the claim struct.

When terminate is called, you'd `close(c.stopChan)`, this loop could have a `case <- c.stopChan` in the select statement. This would allow you to unblock this goroutine as soon as terminate is called.

I find this a bit easier to reason about, the only caveat is you have to make sure multiple threads don't try to close stopChan at the same time.